File: week_1/code/same_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        """
        Given the roots of two binary trees p and q, write a function to check if they are the same or not.
        Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
        """
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False
            
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

# Helper to build tree from level order list
def buildTree(arr):
    if not arr: return None
    root = TreeNode(arr[0])
    queue = [root]
    i = 1
    while queue and i < len(arr):
        node = queue.pop(0)
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    return root

# Test cases
if __name__ == "__main__":
    sol = Solution()
    
    # Test 1: p = [1,2,3], q = [1,2,3]
    p1 = buildTree([1,2,3])
    q1 = buildTree([1,2,3])
    print(f"Test 1: {sol.isSameTree(p1, q1)} (Expected: True)")

    # Test 2: p = [1,2], q = [1,null,2]
    p2 = buildTree([1,2])
    q2 = buildTree([1,None,2])
    print(f"Test 2: {sol.isSameTree(p2, q2)} (Expected: False)")
    
    # Test 3: p = [1,2,1], q = [1,1,2]
    p3 = buildTree([1,2,1])
    q3 = buildTree([1,1,2])
    print(f"Test 3: {sol.isSameTree(p3, q3)} (Expected: False)")

File: week_1/code/same_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function (p, q) {
    if (!p && !q) return true;
    if (!p || !q || p.val !== q.val) return false;

    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
};

// Helper to build tree from array
function buildTree(arr) {
    if (!arr.length) return null;
    let root = new TreeNode(arr[0]);
    let queue = [root];
    let i = 1;
    while (queue.length > 0 && i < arr.length) {
        let node = queue.shift();
        if (i < arr.length && arr[i] !== null) {
            node.left = new TreeNode(arr[i]);
            queue.push(node.left);
        }
        i++;
        if (i < arr.length && arr[i] !== null) {
            node.right = new TreeNode(arr[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

// Test cases
// Test 1: p = [1,2,3], q = [1,2,3]
let p1 = buildTree([1, 2, 3]);
let q1 = buildTree([1, 2, 3]);
console.log("Test 1:", isSameTree(p1, q1), "(Expected: true)");

// Test 2: p = [1,2], q = [1,null,2]
let p2 = buildTree([1, 2]);
let q2 = buildTree([1, null, 2]);
console.log("Test 2:", isSameTree(p2, q2), "(Expected: false)");

// Test 3: p = [1,2,1], q = [1,1,2]
let p3 = buildTree([1, 2, 1]);
let q3 = buildTree([1, 1, 2]);
console.log("Test 3:", isSameTree(p3, q3), "(Expected: false)");
